test_conformance/spirv_new/test_op_spec_constant.cpp

```diff
@@ -0,0 +1,182 @@
+                              device_results.data(), 0, NULL, NULL);
+    SPIRV_CHECK_ERROR(err, "Failed to copy from numbers_buffer");
+
+    for (int i = 0; i < num; i++)
+    {
+        if (device_results[i] != references[i])
+        {
+            log_error(
+                "Values do not match in index %d expected %d obtained %d\n", i,
+                references[i], device_results[i]);
+            return TEST_FAIL;
+        }
+    }
+    return TEST_PASS;
+}
+
+
+#define TEST_SPEC_CONSTANT(NAME, type, value, spec_constant_value)             \
+    TEST_SPIRV_FUNC_VERSION(op_spec_constant_##NAME##_simple, Version(2, 2))   \
+    {                                                                          \
+        std::vector<type> inputs(1024, (type)value);                           \
+        std::vector<type> references(1024,                                     \
+                                     (type)value + (type)spec_constant_value); \
+        return test_spec_constant(deviceID, context, queue,                    \
+                                  "op_spec_constant_" #NAME "_simple", inputs, \
+                                  references, (type)spec_constant_value);      \
+    }
+
+// type name, type, value init, spec constant value
+TEST_SPEC_CONSTANT(int, cl_int, 20, -11)
+TEST_SPEC_CONSTANT(uint, cl_uint, 25, 43)
+TEST_SPEC_CONSTANT(char, cl_char, -20, -3)
+TEST_SPEC_CONSTANT(uchar, cl_uchar, 19, 4)
+TEST_SPEC_CONSTANT(short, cl_short, -6000, -1000)
+TEST_SPEC_CONSTANT(ushort, cl_ushort, 6000, 3000)
+TEST_SPEC_CONSTANT(long, cl_long, 34359738360L, 2)
+TEST_SPEC_CONSTANT(ulong, cl_ulong, 9223372036854775000UL, 200)
+TEST_SPEC_CONSTANT(float, cl_float, 1.5, -3.7)
+TEST_SPEC_CONSTANT(half, cl_half, 1, 2)
+TEST_SPEC_CONSTANT(double, cl_double, 14534.53453, 1.53453)
+
+// Boolean tests
+// documenation: 'If a specialization constant is a boolean
+// constant, spec_value should be a pointer to a cl_uchar value'
+
+TEST_SPIRV_FUNC(op_spec_constant_true_simple)
+{
+    // 1-st ndrange use default spec const (true) value = value + 1 (first check
+    // verifies that values are different) 2-nd ndrange sets spec const false so
+    // value = value - 1
+    cl_uchar value = (cl_uchar)7;
+    std::vector<cl_uchar> inputs(1024, value);
+    std::vector<cl_uchar> references(1024, value);
+    return test_spec_constant<cl_uchar>(deviceID, context, queue,
+                                        "op_spec_constant_true_simple", inputs,
+                                        references, 0);
+}
+
+TEST_SPIRV_FUNC(op_spec_constant_false_simple)
+{
+    // 1-st ndrange use default spec const (false) value = value - 1 (first
+    // check verifies that values are different) 2-nd ndrange sets spec const
+    // true so value = value + 1
+    cl_uchar value = (cl_uchar)8;
+    std::vector<cl_uchar> inputs(1024, value);
+    std::vector<cl_uchar> references(1024, value);
```

Comment:
Not saying this needs to change but: why is this a vector? This will run 1024 times the same test. I'm not sure there's a lot of value in that. We could have much simpler SPIR-V if we changed these tests to operate on a single value.

Reply:
From validation experience more is better then less :). If you insist on change I could prepare.